src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,192 @@
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == 0) {
+            if (first == null) {
+                first = newNode;
+                last = newNode;
+            } else {
+                newNode.next = first;
+                first.prev = newNode;
+                first = newNode;
+            }
+        } else if (index == size) {
+            if (last == null) {
+                first = newNode;
+                last = newNode;
+            } else {
+                newNode.prev = last;
+                last.next = newNode;
+                last = newNode;
+            }
+        } else {
+            Node<T> current = first;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            Node<T> prevNode = current.prev;
+            newNode.next = current;
+            newNode.prev = prevNode;
+            prevNode.next = newNode;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndexFits(index);
+        Node<T> current = first;
+        for (int i = 0; i < size; i++) {
+            if (i == index) {
+                return current.item;
+            }
+            current = current.next;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
+        checkIndexFits(index);
+        Node<T> current = first;
+        for (int i = 0; i < size; i++) {
+            if (i == index) {
+                T result = current.item;
+                current.item = value;
+                return result;
+            }
+            current = current.next;
+        }
         return null;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexFits(index);
+        Node<T> toRemove;
+        if (index == 0) {
+            toRemove = first;
+            first = first.next;
+            if (first != null) {
+                first.prev = null;
+            } else {
+                last = null;
+            }
+        } else if (index == size - 1) {
+            toRemove = last;
+            last = last.prev;
+            if (last != null) {
+                last.next = null;
+            } else {
+                first = null;
+            }
+        } else {
+            Node<T> current = first;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            toRemove = current;
+            Node<T> prevNode = current.prev;
+            Node<T> nextNode = current.next;
+
+            prevNode.next = nextNode;
+            nextNode.prev = prevNode;
+        }
+
+        size--;
+        return toRemove.item;
```

Comment:
The logic for unlinking a node is repeated; consider creating an 'unlink' method that can be used in both 'remove' methods.